docs - troubleshooting guide: load from h2 (#17209)
metabase/metabase: docs/operations-guide/migrating-from-h2.md

# Migrating from using the H2 database to MySQL or Postgres

> If you decide to use the default application database (H2) when you initially start using Metabase, but later decide that you'd like to switch to a more production-ready database such as MySQL or Postgres, you're in the right place.

Reviewer: ```suggestion
If you decide to use the default application database (H2) when you initially start using Metabase, but later decide that you'd like to switch to a high-availability database such as MySQL or Postgres, you're in the right place.
```

Reviewer 2: Postgres/MySQL are not HA unless they are configured as such.

## Before you migrate

> - Avoid upgrading and migrating at the same time, since it can cause problems with one of database schemas not matching.

Reviewer: ```suggestion
- Avoid upgrading and migrating at the same time, since it can cause problems if one of database schemas doesn't match.
```

> - You must be able to connect to the target MySQL or Postgres database in whatever environment you're running this migration command in. So, if you are attempting to move the data to a cloud database, make sure you can connect to that database.

Reviewer: Q: do we only support MySQL and Postgres as replacements for H2? This makes it sound as if that's so.

Reviewer 2: Yes, only Postgres and MySQL/MariaDB are supported as appdb - besides H2.

## How to migrate 

> Metabase provides a custom migration command for upgrading H2 application database files by copying their data to a new database. Here's what you'll want to do:

Reviewer: ```suggestion
Metabase provides a custom migration command for upgrading H2 application database files by copying their data to a new database. Here's what you'll do:
```

export MB_DB_USER=<username>
export MB_DB_PASS=<password>
export MB_DB_HOST=localhost
java -jar metabase.jar load-from-h2 /path/to/metabase.db # do not include .mv.db
```

> It is expected that you will run the command against a brand-new (empty!) database; Metabase will handle all of the work of creating the database schema and migrating the data for you.

Reviewer: ```suggestion
Metabase expects that you will run the command against a brand-new (empty) database; it will create the database schema and migrate the data for you.
```

### PostgreSQL notes

-  Minimum version is PostgreSQL 9.4, since the code that handles these migrations uses a command that is only available in version 9.4 or newer.

## MySQL/MariaDB notes

- MySQL minimum recommended version is 5.7.7.
- MariaDB minimum recommended version is 10.2.2.
- And the following database settings are required, which is the default in the above recommended versions: `utf8mb4_unicode_ci` collation, `utf8mb4` character set, and `innodb_large_prefix=ON`.

### Troubleshooting

> If you get stuck, check out [Error when loading application database from H2](../troubleshooting-guide/loading-from-h2.md).

Reviewer: What does "get stuck" mean in this context? I.e, what are the symptoms someone should look for?

> If you get stuck, check out [Error when loading application database from H2](../troubleshooting-guide/loading-from-h2.md).

Author: ```suggestion
If you get an error, check out [Error when loading application database from H2](../troubleshooting-guide/loading-from-h2.md).
```

export MB_DB_USER=<username>
export MB_DB_PASS=<password>
export MB_DB_HOST=localhost

> java -jar metabase.jar load-from-h2 /path/to/metabase.db.mv.db 

Reviewer 2: ```suggestion
java -jar metabase.jar load-from-h2 /path/to/metabase.db # do not include .mv.db
```

```

It is expected that you will run the command against a brand-new (empty!) database; Metabase will handle all of the work of creating the database schema and migrating the data for you.

### PostgreSQL notes

-  Minimum version is PostgreSQL 9.4, since the code that handles these migrations uses a command that is only available in version 9.4 or newer.

> ## MySQL notes
> 
> - MySQL minimum recommended version is 5.7.7.
> 
> ### MariaDB notes
> 
> - MariaDB minimum recommended version is 10.2.2, while the following is required: `utf8mb4_unicode_ci` collation, `utf8mb4` character set, and `innodb_large_prefix=ON`.

Reviewer 2: ```suggestion
## MySQL/MariaDB notes

- MySQL minimum recommended version is 5.7.7.
- MariaDB minimum recommended version is 10.2.2.
- And the following database settings are required, which is the default in the above recommended versions: `utf8mb4_unicode_ci` collation, `utf8mb4` character set, and `innodb_large_prefix=ON`.
```
